File: mac/McController/Core/Diag/ConnectionStats.swift
import Foundation

/// Lock-light counters shared between the network layer and the UI.
/// Counters are monotonic; the UI samples them periodically. Mirrors
/// `ConnectionStats.cs`.
///
/// `@MainActor` is intentionally not applied — counter updates fire from
/// the network read loops. Synchronization is via an internal queue +
/// `OSAllocatedUnfairLock` for the small mutable state.
final class ConnectionStats: @unchecked Sendable {

    private let lock = NSLock()

    private var _connected: Bool = false
    private var _clientEndpoint: String?
    private var _mode: String?
    private var _joystickCount: UInt64 = 0
    private var _lookCount: UInt64 = 0
    private var _buttonCount: UInt64 = 0
    private var _udpDropped: UInt64 = 0
    private var _rttSamples: [Int] = []
    private let rttWindow = 60

    var connected: Bool {
        get { lock.withLock { _connected } }
        set { lock.withLock { _connected = newValue } }
    }

    var clientEndpoint: String? {
        get { lock.withLock { _clientEndpoint } }
        set { lock.withLock { _clientEndpoint = newValue } }
    }

    /// Display label, e.g. "WiFi (TCP+UDP)" or "USB (TCP only)".
    var mode: String? {
        get { lock.withLock { _mode } }
        set { lock.withLock { _mode = newValue } }
    }

    var joystickCount: UInt64 { lock.withLock { _joystickCount } }
    var lookCount: UInt64 { lock.withLock { _lookCount } }
    var buttonCount: UInt64 { lock.withLock { _buttonCount } }
    var udpDropped: UInt64 { lock.withLock { _udpDropped } }

    func incrementJoystick() { lock.withLock { _joystickCount &+= 1 } }
    func incrementLook()     { lock.withLock { _lookCount &+= 1 } }
    func incrementButton()   { lock.withLock { _buttonCount &+= 1 } }
    func incrementUdpDropped() { lock.withLock { _udpDropped &+= 1 } }

    func recordRtt(ms: Int) {
        lock.withLock {
            _rttSamples.append(ms)
            if _rttSamples.count > rttWindow {
                _rttSamples.removeFirst(_rttSamples.count - rttWindow)
            }
        }
    }

    func rttPercentiles() -> (p50: Int, p99: Int) {
        let sorted: [Int] = lock.withLock {
            if _rttSamples.isEmpty { return [] }
            return _rttSamples.sorted()
        }
        if sorted.isEmpty { return (0, 0) }
        let p50 = sorted[sorted.count / 2]
        let p99Idx = min(Int(floor(Double(sorted.count) * 0.99)), sorted.count - 1)
        return (p50, sorted[p99Idx])
    }

    func onDisconnect() {
        lock.withLock {
            _connected = false
            _clientEndpoint = nil
            _mode = nil
            _joystickCount = 0
            _lookCount = 0
            _buttonCount = 0
            _udpDropped = 0
            _rttSamples.removeAll()
        }
    }
}

private extension NSLock {
    func withLock<T>(_ body: () -> T) -> T {
        lock(); defer { unlock() }
        return body()
    }
}

File: pc/McController.Core/Diag/ConnectionStats.cs
namespace McController.Core.Diag;

/// <summary>
/// Lock-light counters and connection state shared between the network layer
/// and the UI. Counters are monotonic; the UI samples them periodically and
/// computes per-second rates locally.
/// </summary>
public sealed class ConnectionStats
{
    private volatile bool _connected;
    private volatile string? _clientEndpoint;
    private volatile string? _mode;

    private long _joystickCount;
    private long _lookCount;
    private long _buttonCount;
    private long _udpDropped;

    public bool Connected
    {
        get => _connected;
        set => _connected = value;
    }

    public string? ClientEndpoint
    {
        get => _clientEndpoint;
        set => _clientEndpoint = value;
    }

    /// <summary>Display label, e.g. "WiFi (TCP+UDP)" or "USB (TCP only)".</summary>
    public string? Mode
    {
        get => _mode;
        set => _mode = value;
    }

    public long JoystickCount => Interlocked.Read(ref _joystickCount);
    public long LookCount => Interlocked.Read(ref _lookCount);
    public long ButtonCount => Interlocked.Read(ref _buttonCount);
    public long UdpDropped => Interlocked.Read(ref _udpDropped);

    public void IncrementJoystick() => Interlocked.Increment(ref _joystickCount);
    public void IncrementLook() => Interlocked.Increment(ref _lookCount);
    public void IncrementButton() => Interlocked.Increment(ref _buttonCount);
    public void IncrementUdpDropped() => Interlocked.Increment(ref _udpDropped);

    private readonly Queue<int> _rttSamples = new();
    private readonly object _rttLock = new();
    private const int RttWindowSize = 60;  // ~60 1Hz samples = 1 minute window

    public void RecordRtt(int rttMs)
    {
        lock (_rttLock)
        {
            _rttSamples.Enqueue(rttMs);
            while (_rttSamples.Count > RttWindowSize) _rttSamples.Dequeue();
        }
    }

    public (int p50, int p99) RttPercentiles()
    {
        int[] sorted;
        lock (_rttLock)
        {
            if (_rttSamples.Count == 0) return (0, 0);
            sorted = _rttSamples.ToArray();
        }
        Array.Sort(sorted);
        var p50 = sorted[sorted.Length / 2];
        var p99Idx = Math.Min((int)Math.Floor(sorted.Length * 0.99), sorted.Length - 1);
        var p99 = sorted[p99Idx];
        return (p50, p99);
    }

    public void OnDisconnect()
    {
        Connected = false;
        ClientEndpoint = null;
        Mode = null;
        Interlocked.Exchange(ref _joystickCount, 0);
        Interlocked.Exchange(ref _lookCount, 0);
        Interlocked.Exchange(ref _buttonCount, 0);
        Interlocked.Exchange(ref _udpDropped, 0);
        lock (_rttLock) _rttSamples.Clear();
    }
}
